crate/utils/src/crypto/curve_25519/encryption_decryption.rs

```diff
@@ -1,49 +1,38 @@
-use std::sync::{Arc, Mutex};
-
 use cloudproof::reexport::crypto_core::{
-    asymmetric_crypto::{
-        curve25519::{X25519KeyPair, X25519PrivateKey, X25519PublicKey},
-        DhKeyPair,
-    },
-    bytes_ser_de::Serializable,
-    reexport::rand_core::SeedableRng,
-    CsRng,
+    bytes_ser_de::Serializable, reexport::rand_core::SeedableRng, CsRng, Ecies,
+    EciesX25519XChaCha20, FixedSizeCBytes, X25519PrivateKey, X25519PublicKey,
 };
 use cosmian_kmip::kmip::{
     kmip_objects::Object,
     kmip_operations::{Decrypt, DecryptResponse, DecryptedData, Encrypt, EncryptResponse},
 };
 use tracing::{debug, trace};
 
-use crate::crypto::{
-    ecies::{ecies_decrypt, ecies_encrypt},
-    error::{result::CryptoResultHelper, CryptoError},
+use crate::{
+    crypto::error::{result::CryptoResultHelper, CryptoError},
+    DecryptionSystem, EncryptionSystem,
 };
-// use super::user_key::unwrap_user_decryption_key_object;
-use crate::{DecryptionSystem, EncryptionSystem};
 
 /// Encrypt a single block of data using an hybrid encryption mode
 /// Cannot be used as a stream cipher
 pub struct EciesEncryption {
-    rng: Arc<Mutex<CsRng>>,
     public_key_uid: String,
     public_key: X25519PublicKey,
 }
 
-/// Maximum clear text size that can be safely encrypted with AES GCM (using a single random nonce)
-pub const MAX_CLEAR_TEXT_SIZE: usize = 1_usize << 30;
+/// x25519 key length
+pub const X25519_PUBLIC_KEY_LENGTH: usize = 32;
+pub const X25519_PRIVATE_KEY_LENGTH: usize = 32;
 
 impl EciesEncryption {
     pub fn instantiate(public_key_uid: &str, public_key: &Object) -> Result<Self, CryptoError> {
-        let rng = CsRng::from_entropy();
-
-        let public_key_bytes = public_key.key_block()?.key_bytes()?;
-        let public_key = X25519PublicKey::try_from_bytes(&public_key_bytes)?;
+        let public_key_bytes: [u8; X25519_PUBLIC_KEY_LENGTH] =
+            public_key.key_block()?.key_bytes()?.as_slice().try_into()?;
+        let public_key = X25519PublicKey::try_from_bytes(public_key_bytes)?;
 
         trace!("Instantiated hybrid ECIES encipher for public key id: {public_key_uid}");
 
         Ok(Self {
-            rng: Arc::new(Mutex::new(rng)),
```

Comment:
You are losing performances by moving the RNG instantiation to the `encrypt` function. Is there a good reason for that?

Reply:
reverted